syncer/syncer.go

```diff
@@ -813,21 +826,10 @@ func (s *Syncer) Connect(ctx context.Context, addr string) (*Peer, error) {
 		ConnAddr: conn.RemoteAddr().String(),
 		Inbound:  false,
 	}
-	go func() {
-		done, err := s.tg.Add()
-		if err != nil {
-			return
-		}
-		s.runPeer(p)
-		done()
-	}()
+	if _, err := s.runPeer(p); err != nil {
+		return nil, fmt.Errorf("failed to add peer: %w", err)
+	}
 
-	// runPeer does this too, but doing it outside the goroutine prevents a race
```

Comment:
My PR re-introduces this race. I wanted to draw attention to this, I did that because I don't see why it's that much of an issue. But also because fixing that would require `runPeer` to signal the peer was added, as well as be able to execute it in a blocking fashion. Which we could do return two channels but since it's already a bit of anti-pattern I didn't dare pushing that into review 😅

Reply:
What I don't get is why introduce this anti-pattern instead of breaking up `runPeer` into 2 methods to avoid the channel?

1. `addPeer` -> calls `pm.AddPeer`, `pm.UpdatePeerInfo` and adds the peer to `s.peers` so basically all the things we want to wait for
2. `runPeer` -> spawns the goroutine to handle the peer and handle the asynchronous stuff

Then you can call `addPeer` and check for an error indicating whether the peer was added or not before launching `runPeer` in a goroutine. So you don't need to add the peer here anymore.

I think than you can also keep most of the remaining codebase untouched assuming these lines were the culprit causing the issue.

Reply:
Well because I personally prefer the channel to be able to span the goroutine that handles the removal as close as possible to where we add the peer to `s.peers` 🤷🏻‍♂️ . That was before I made this comment though... I'll update the PR